streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java

```diff
@@ -111,5 +120,75 @@ public R apply(T2 value2, T1 value1) {
                 .enableCaching();
     }
 
+    static <K, V, VR> ValueMapperWithKey<K, V, VR> withKey(final ValueMapper<V, VR> valueMapper) {
+        Objects.requireNonNull(valueMapper, "valueMapper can't be null");
+        return new ValueMapperWithKey<K, V, VR>() {
+            @Override
+            public VR apply(final K readOnlyKey, final V value) {
+                return valueMapper.apply(value);
+            }
+        };
+    }
 
+    static <K, V, VR> InternalValueTransformerWithKeySupplier<K, V, VR> toInternalValueTransformerSupplier(final ValueTransformerSupplier<V, VR> valueTransformerSupplier) {
+        Objects.requireNonNull(valueTransformerSupplier, "valueTransformerSupplier can't be null");
+        final ValueTransformer<V, VR> valueTransformer = valueTransformerSupplier.get();
+        return new InternalValueTransformerWithKeySupplier<K, V, VR>() {
+            @Override
+            public InternalValueTransformerWithKey<K, V, VR> get() {
+                return new InternalValueTransformerWithKey<K, V, VR>() {
+                    @Override
+                    public VR punctuate(final long timestamp) {
+                        return valueTransformer.punctuate(timestamp);
+                    }
+
+                    @Override
+                    public void init(final ProcessorContext context) {
+                        valueTransformer.init(context);
+                    }
+
+                    @Override
+                    public VR transform(final K readOnlyKey, final V value) {
+                        return valueTransformer.transform(value);
+                    }
+
+                    @Override
+                    public void close() {
+                        valueTransformer.close();
+                    }
+                };
+            }
+        };
+    }
+
+    static <K, V, VR> InternalValueTransformerWithKeySupplier<K, V, VR> toInternalValueTransformerSupplier(final ValueTransformerWithKeySupplier<K, V, VR> valueTransformerWithKeySupplier) {
+        Objects.requireNonNull(valueTransformerWithKeySupplier, "valueTransformerSupplier can't be null");
+        final ValueTransformerWithKey<K, V, VR> valueTransformerWithKey = valueTransformerWithKeySupplier.get();
+        return new InternalValueTransformerWithKeySupplier<K, V, VR>() {
+            @Override
+            public InternalValueTransformerWithKey<K, V, VR> get() {
+                return new InternalValueTransformerWithKey<K, V, VR>() {
+                    @Override
+                    public VR punctuate(final long timestamp) {
+                        return null;
```

Comment:
I am wondering if we should `throw` here as this should never be called. Or do I miss anything?